monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/Interleave2Observable.scala

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (c) 2014-2016 by its authors. Some rights reserved.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.reactive.internal.builders
+
+
+import monix.execution.Ack.{Stop, Continue}
+import monix.execution.cancelables.CompositeCancelable
+import monix.execution.{Ack, Cancelable}
+import monix.reactive.Observable
+import monix.reactive.observers.Subscriber
+
+import scala.concurrent.{Future, Promise}
+import scala.util.Success
+
+private[reactive] final class Interleave2Observable[+A]
+  (obsA1: Observable[A], obsA2: Observable[A]) extends Observable[A] { self ⇒
+
+  val idx = new java.util.concurrent.atomic.AtomicInteger(0)
+
+  def unsafeSubscribeFn(out: Subscriber[A]): Cancelable = {
+    import out.scheduler
+
+    // MUST BE synchronized by `self`
+    var isDone = false
+    // MUST BE synchronized by `self`
+    var lastAck = Continue : Future[Ack]
```

Comment:
We have too many `lastAck` stuff. Rename this to `downstreamAck` (being the acknowledgement that we receive from the subscriber when we actually send something).